clang/lib/Sema/SemaInit.cpp

```diff
@@ -9102,6 +9102,15 @@ bool InitializationSequence::Diagnose(Sema &S,
 
   case FK_ConversionFailed: {
     QualType FromType = OnlyArg->getType();
+    // __amdgpu_feature_predicate_t can be explicitly cast to the logical op
+    // type, although this is almost always an error and we advise against it
```

Comment:
```suggestion
    // __amdgpu_feature_predicate_t can be explicitly cast to the logical op
    // type, although this is almost always an error and we advise against it.
```
I'm a bit confused; the comment says it's allowed but implies we should at least warn on it. But we're emitting an error diagnostic for that?

Reply:
The error is trying to do this without a cast, which is what is being tested (this is the same behaviour one gets around a type that has an `explicit` conversion operator for bool). The error message is trying to inform the user that if they REALLY want to do this, they can, but they have to explicitly cast via C-cast or C++ `static_cast`, however they probably should not as it is almost always going to lead to an error. I am happy to reword the error message / add a warning on casting from `__amdgpu_feature_predicate_t` to `bool` / `_Bool`.

Reply:
> Ah, I see the logic now, I just didn't read hard enough. :-D
> 
> So why would the explicit cast almost always lead to an error given that we're doing the cast implicitly for them anyway?

Context:) Since we're modelling this after a type with an `explicit operator bool()`, which is not regular (no default, copy or move ctors) only contextual conversions are valid i.e. (off the top of my head, I'd have to look up the exact reference in the standard):
- controlling expression for `if`, `while`, `for`;
- first operand of the ternary / conditional operator `?`;
- operands for guilt-in logical operators;
- some `constexpr` / `consteval` cases that do not apply.
So, we do want `if (__builtin_amdgcn_processor_is("gfx900")` to just work as typed, as this is an intended usage pattern (user checks for a particular target, then does something that is target specific on the `true` branch). 

However, if your expression is not contextually convertible to `bool`/`_Bool`, you are probably walking into trouble by doing something potentially nefarious like this:

```cpp
void foo(const std::vector<bool>& ps) { 
  // do stuff based on indexing into the predicate vector, e.g.
  // assume that the 0th element holds gfx900, the 1st holds 
  // gfx901 etc.
}

void bar() {
  vector<bool> ps;
  ps.push_back(__builtin_amdgcn_processor_is("gfx900");
  ps.push_back(__builtin_amdgcn_processor_is("gfx901");
}
```

which might make a lot of internal sense for the client app, and looks pretty harmless, but is
not something we can support in the limit (once it's buried under  5 additional layers of indirection, at `-O0` etc.), and which is likely to end up a bug farm. These really do have sharp edges so the intention is to be defensive by default and ensure that it's very hard to step on a landmine.

Reply:
> or abstract (amdgcnspirv), i.e. we only know the concrete target at run time

Okay, I think this was the key bit I was missing. I had the impression that we knew this information up front at compile time. At runtime, is the answer consistent? e.g., is this something we know at program startup and it won't change as the program runs? Or is this information more ephemeral than that?

Reply:
It is consistent. The run time finalisation is based on which targets are present in a machine / environment, so at the point where we run the predicate expansion pass we know exactly what the target is. Even in the rather exotic case of hot-plugging an accelerator that was not part of the initial set, what would happen is that the language (e.g. HIP) runtime would try to dispatch, look up code for the newcomer, see that there is none and then finalise for that target. So, overall, we either have the target specific code, or we can generate/finalise it, but there's no ambiguity at run time.

Reply:
Let me see if I'm on the same page now. The backend will generate code for gfx900 and the programmer will guard that block of code with `if (__builtin_amdgcn_processor_is("gfx900"))`. So if the predicate and the block it controls become disjointed somehow, the user can get incorrect behavior at runtime. Am I on the right track?

Reply:
Not really, although cleanliness, just like beauty, is in the eye of the beholder:) The feature builtins work locally, and naturally controls what might be a single ASM block, or builtin invocation, or mmio access etc. Going with target_clones has a number of implementation problems (it'd require a bunch of other things that are neither there nor impending for AMDGPU / SPIRV to materialise first), it forces outlining / adopting a different software development paradigm, and one that is fairly spammy (do I duplicate the full function (10s / 100s of lines, possibly) just because I might have some target specific behaviour? Do I start sharding out things that are target specific essentially creating a parallel world of `target_clones` based builtins, even though this code might be mature and gnarly etc.). The feature builtins allow a linear, inline transform from:

```cpp
#if defined(__gfx900__)
    __builtin_amdgcn_only_on_gfx900();
#endif
```

to

```cpp
if (__builtin_amdgcn_processor_is("gfx900"))
    __builtin_amdgcn_only_on_gfx900();
```
which is easy to teach / adopt and matches what folks are already doing, and just works as is, both on concrete, compile-time finalised targets, and on abstract run-time finalised ones. Also please consider that we are focusing on the `processor_is` case, however the preferable, more adaptable solution would be to check for the availability of a builtin, rather than an architecture, and that'd be even more spammy (and would make even more sense inline). 

Furthermore, is the current formulation particularly novel? It's simply a builtin that returns an irregular type with an explicit conversion operator for `bool`. Holding it right in this case matches what people already do; none of the exciting indirections we are discussing here would be at all possible today. Finally, we already have e.g. [__builtin_available](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available) or `__nvvm_reflect`, which tackle similar problems via similar means, and can be misused in similar ways.

Reply:
> The feature builtins allow a linear, inline transform from ... to ... which is easy to teach / adopt and matches what folks are already doing, and just works as is, both on concrete, compile-time finalised targets, and on abstract run-time finalised ones. 

They work as-is so long as you hold them right. What I think is not easy to teach or adopt is the fact that it's tied so tightly to optimizer behavior in surprising ways. The fact that these two are not necessarily equivalent (esp as the logic becomes more complex) is a pretty reasonable concern to have:
```
if (__builtin_amdgcn_processor_is("gfx900"))
  do_stuff();

bool b = (bool)__builtin_amdgcn_processor_is("gfx900");
if (b)
  do_stuff();
```
because that's not intuitive. What's worse, debugging this when your intuition is wrong will be *incredibly* difficult because the behavior in a -O0 build is very likely going to do the right thing, so you end up debugging optimized code instead.

> Furthermore, is the current formulation particularly novel?

The shape of it isn't novel (a feature test which takes an argument and returns something truthy), which is actually why I'm pushing back so hard. This looks and feels like `__has_builtin` or other feature testing macros. AFAIK, those aren't tied to optimizer behavior so tightly. I think the default expectation is that you should be able to query the processor information at any point you want, store the results anywhere you want, and use them later with the expected semantics; any other behavior starts to feel like a miscompile.

As a counter-example, consider `__has_builtin`:
```
bool has_builtin = __has_builtin(__builtin_whatever);
// ...later...
if (has_builtin)
  __builtin_whatever();
```
if the call to `__has_builtin` returned false and we still ended up calling `__builtin_whatever`, users would be baffled as to why. And if it returned `true` and we didn't call the builtin, users would be just as baffled. So what I'm struggling to understand is why these situations are materially different from each other.

Reply:
> I suspect that the sense that something is tied to optimiser behaviour is due to my reply above, which perhaps was insufficiently clear - apologies.

No worries, this is complex stuff! I appreciate your willingness to talk me through it. :-)

> The __has_builtin counter-example actually does not work and cannot work, please see: https://gcc.godbolt.org/z/7G5Y1d85b.

I cannot imagine a situation in which this isn't indicative of a bug, but perhaps this situation is the same one that necessitated [this PR](https://github.com/llvm/llvm-project/pull/126324#issuecomment-2706655366) which eventually concluded that we should change the behavior of `__has_builtin` rather than introduce a new builtin.

> furthermore, the ...later... bit is pretty important: what happens on that path?

Anything in the world besides changing the value of `has_builtin` to something other than what `__has_builtin` returned.

>  if you do any of those, your distant has_builtin variable no longer reflects the predicate, which is the issue;

Why is that an issue? If the variable no longer reflects the predicate, that's not on the compiler to figure out how to deal with, that's "play silly games, win silly prizes".

Backing up a step.. my expectation is that this eventually lowers down to a test and jump which jumps past the target code if the test fails. e.g.,
```
  %0 = load i8, ptr %b, align 1
  %loadedv = trunc i8 %0 to i1
  br i1 %loadedv, label %if.then, label %if.end

if.then:
  # the target-specific instructions live here
  br label %if.end

if.end:
  ret void
```
So we'd be generating instructions for the target which may be invalid if the test lies. If something did change that value so it no longer represents the predicate, I think that's UB (and we could help users catch that UB via a sanitizer check if we wanted to, rather than try to make the backend have to try to figure it out at compile time).

> if for a chain from point of cast to final point of use folding fails (because you passed your value to an
opaque function, modified it based on a run time value etc.), you get an error and a diagnostic.

I was thinking you would not get a diagnostic; you'd get the behavior you asked for, which may be utter nonsense.

Am I missing something still? If so, maybe it would be quicker for us to hop in a telecon call? I'm going to be out of the office until Monday, but I'm happy to meet with you if that's more productive.

Reply:
> > The __has_builtin counter-example actually does not work and cannot work, please see: https://gcc.godbolt.org/z/7G5Y1d85b.
> 
> I cannot imagine a situation in which this isn't indicative of a bug, but perhaps this situation is the same one that necessitated [this PR](https://github.com/llvm/llvm-project/pull/126324#issuecomment-2706655366) which eventually concluded that we should change the behavior of `__has_builtin` rather than introduce a new builtin.
> 

This is not actually a bug, it's intended behaviour. To obtain what you expect the `b` would have to be `constexpr`, and then the `if` itself would have to be `if constexpr`. Otherwise there's no binding commitment to evaluate this at compile time (and, in effect, if this gets trivially evaluated / removed in the FE, it induces dependence on optimisation level).

> > furthermore, the ...later... bit is pretty important: what happens on that path?
> 
> Anything in the world besides changing the value of `has_builtin` to something other than what `__has_builtin` returned.
> 
> > if you do any of those, your distant has_builtin variable no longer reflects the predicate, which is the issue;
> 
> Why is that an issue? If the variable no longer reflects the predicate, that's not on the compiler to figure out how to deal with, that's "play silly games, win silly prizes".
>

It is a difficult conversation to have and not exactly what users want to hear, so making it as hard as possible to end up in an exchange where you have to say "welp, that was a silly game" cannot hurt. If anything, it's compassionate behaviour!

> Backing up a step.. my expectation is that this eventually lowers down to a test and jump which jumps past the target code if the test fails. e.g.,
> 
> ```
>   %0 = load i8, ptr %b, align 1
>   %loadedv = trunc i8 %0 to i1
>   br i1 %loadedv, label %if.then, label %if.end
> 
> if.then:
>   # the target-specific instructions live here
>   br label %if.end
> 
> if.end:
>   ret void
> ```
> 
> So we'd be generating instructions for the target which may be invalid if the test lies. If something did change that value so it no longer represents the predicate, I think that's UB (and we could help users catch that UB via a sanitizer check if we wanted to, rather than try to make the backend have to try to figure it out at compile time).
> 

This cannot work reliably (e.g. there are instructions that would simply fail at ISEL, and a run time jump doesn't mean that you do not lower to ISA the jumped around block), and introducing dependence on sanitizers seems not ideal. Furthermore, a run time jump isn't free, which is a concern for us, and we also already have a mechanism for that case (`__attribute__((target))`).  Note that these can also control e.g. resource allocation, so actually generating both might lead to arbitrary exhaustion of a limited resource, and spurious compilation failures, consider e.g. (I'll use CUDA/HIP syntax):

```cpp
// This is a bit odd, and technically a race because multiple lanes write to shared_buf
void foo() {
  __shared__ int* shared_buf;
  if (__builtin_amdgcn_processor_is("gfx950") {
    __shared__ int buf[70 * 1024];
    shared_buf = buf;
  } else {
    __shared__ int buf[60 * 1024];
    shared_buf = buf;
  }

  __syncthreads();
  // use shared_buf
```

If we tried to lower that we'd exhaust LDS, and spuriously fail to compile. This would have originated from perfectly valid uses of `#if defined(__gfx950__) #else`. We'd like these to work, so we must unambiguously do the fold ourselves.

> > if for a chain from point of cast to final point of use folding fails (because you passed your value to an
> > opaque function, modified it based on a run time value etc.), you get an error and a diagnostic.
> 
> I was thinking you would not get a diagnostic; you'd get the behavior you asked for, which may be utter nonsense.
> 

One of the difficulties here (ignoring that the utter nonsense behaviour at run time might be nasal demons - GPUs aren't always as polite as to issue a `SIGILL` and graciously die:)) is that not all constructs / IR sequences / ASM uses lower into ISA, so what the user is more likely to get is an ICE with an error that makes no sense unless they work on LLVM. That's fairly grim user experience, IMHO, and one that we have the ability to prevent.

> Am I missing something still? If so, maybe it would be quicker for us to hop in a telecon call? I'm going to be out of the office until Monday, but I'm happy to meet with you if that's more productive.

I would be absolutely happy to if you think it'd help. I regret not coming to the Sofia meeting, we could've probably sorted this out directly with a laptop:)

Reply:
> > > The __has_builtin counter-example actually does not work and cannot work, please see: https://gcc.godbolt.org/z/7G5Y1d85b.
> > 
> > 
> > I cannot imagine a situation in which this isn't indicative of a bug, but perhaps this situation is the same one that necessitated [this PR](https://github.com/llvm/llvm-project/pull/126324#issuecomment-2706655366) which eventually concluded that we should change the behavior of `__has_builtin` rather than introduce a new builtin.
> 
> This is not actually a bug, it's intended behaviour. To obtain what you expect the `b` would have to be `constexpr`, and then the `if` itself would have to be `if constexpr`. Otherwise there's no binding commitment to evaluate this at compile time (and, in effect, if this gets trivially evaluated / removed in the FE, it induces dependence on optimisation level).

I... am an idiot. :-D Sorry, I think I must have been braindead when I wrote that because you're exactly correct. Sorry for the noise!

> > Backing up a step.. my expectation is that this eventually lowers down to a test and jump which jumps past the target code if the test fails. e.g.,
> > ```
> >   %0 = load i8, ptr %b, align 1
> >   %loadedv = trunc i8 %0 to i1
> >   br i1 %loadedv, label %if.then, label %if.end
> > 
> > if.then:
> >   # the target-specific instructions live here
> >   br label %if.end
> > 
> > if.end:
> >   ret void
> > ```
> >   
> > So we'd be generating instructions for the target which may be invalid if the test lies. If something did change that value so it no longer represents the predicate, I think that's UB (and we could help users catch that UB via a sanitizer check if we wanted to, rather than try to make the backend have to try to figure it out at compile time).
> 
> This cannot work reliably (e.g. there are instructions that would simply fail at ISEL, and a run time jump doesn't mean that you do not lower to ISA the jumped around block), and introducing dependence on sanitizers seems not ideal. Furthermore, a run time jump isn't free, which is a concern for us, and we also already have a mechanism for that case (`__attribute__((target))`). Note that these can also control e.g. resource allocation, so actually generating both might lead to arbitrary exhaustion of a limited resource, and spurious compilation failures, consider e.g. (I'll use CUDA/HIP syntax):
> 
> ```c++
> // This is a bit odd, and technically a race because multiple lanes write to shared_buf
> void foo() {
>   __shared__ int* shared_buf;
>   if (__builtin_amdgcn_processor_is("gfx950") {
>     __shared__ int buf[70 * 1024];
>     shared_buf = buf;
>   } else {
>     __shared__ int buf[60 * 1024];
>     shared_buf = buf;
>   }
> 
>   __syncthreads();
>   // use shared_buf
> ```
> 
> If we tried to lower that we'd exhaust LDS, and spuriously fail to compile. This would have originated from perfectly valid uses of `#if defined(__gfx950__) #else`. We'd like these to work, so we must unambiguously do the fold ourselves.

Okay, so the situation is different than what I expected. I was unaware this would cause ISEL failures.

> > > if for a chain from point of cast to final point of use folding fails (because you passed your value to an
> > > opaque function, modified it based on a run time value etc.), you get an error and a diagnostic.
> > 
> > 
> > I was thinking you would not get a diagnostic; you'd get the behavior you asked for, which may be utter nonsense.
> 
> One of the difficulties here (ignoring that the utter nonsense behaviour at run time might be nasal demons - GPUs aren't always as polite as to issue a `SIGILL` and graciously die:)) is that not all constructs / IR sequences / ASM uses lower into ISA, so what the user is more likely to get is an ICE with an error that makes no sense unless they work on LLVM. That's fairly grim user experience, IMHO, and one that we have the ability to prevent.

Yeah, we obviously don't want the user experience to be compiler crashes. :-) 

> > Am I missing something still? If so, maybe it would be quicker for us to hop in a telecon call? I'm going to be out of the office until Monday, but I'm happy to meet with you if that's more productive.
> 
> I would be absolutely happy to if you think it'd help. I regret not coming to the Sofia meeting, we could've probably sorted this out directly with a laptop:)

FWIW, I'm still pretty uncomfortable about this design. I keep coming back to this feeling really novel and seeming like it's designed to work around backend issues. If the user did something like this:
```
void func(std::vector<bool> processor_features) {
  if (processor_features[12]) { // SSE3 is allowed
      __asm__ ("do a bunch of sse3 stuff");
  } else {
    // Do slow fallback stuff
  }
}
```
they would reasonably expect that inline assembly to be non-problematic even if sse3 isn't available. But... when I try to play silly games in practice, we assert: https://godbolt.org/z/xc13WhW4W and so maybe I'm just wrong. CC @nikic for more opinions.

As for meeting to discuss, are you free sometime this week? I'm on US East Coast time, what times typically work best for you?

Reply:
From the bottom up, anything but Friday should be good, including today starting from now to now + 6 hours:) I'm in the UK, so the delta is not so large anyway, pick something that fits your schedule and I'll probably be able to make it work.

For your example at the bottom, the ASM is non-problematic in that it goes through. Now substitute it with a builtin that is only there iff SSE3 is available, or try to bind registers from the extended x86_64 set and compile for an x86 target, and it'll go back to failing at compile time. It's that latter part that is problematic even for the user's experience.

I suspect that part of the issue here is that something like X86 hides a lot of this stuff under normal circumstances because folks don't really normally grab for special functionality, or feel the need for it. But if we have a look at the many SIMD extension sets, as well as the attempts at defining various levels of capability (the v1, v2, v3 things) I think the same challenge exists there, it's just not an immediate concern. What we're trying to solve is that whilst it makes perfect sense for a BE, any BE, to bind very tightly to the target, it is sometimes beneficial for the IR coming out of the FE to be generic and usable by many targets, without loss of capability. Without a mechanism as the one here one is either degraded to lowest common denominator capability, or has to play games trying to define capability levels, which generally end up being too coarse.

Also, please note that, in spite of me mentioning x86, at this point we are not proposing this for general use, but rather as a target specific BI, which hopefully reduces risk / contains any perceived novelty to parts where it's already been found to be useful:)